### DISCUSSION TITLE
ZHA in Home Assistant to notify on Zigbee OTA "update availability" and give users UI options to apply specific OTAU update per Zigbee device on demand?

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#951] [ZHA development]
Feature request discussion about ZHA Zigbee OTA "update availability" and UI/GUI options to apply OTAU updates per device "on demand".

https://community.home-assistant.io/t/zha-integration-ui-options-to-apply-ota-firmware-update-otau-per-zigbee-device/438227/

This is obviously not going to be possible unless zigpy get the ability to perform OTA updates on demand, as opposed to in the background.

Back-story; This feature request for ZHA's UI to feature "Update Device Firmware" function for OTA firmware updates of Zigbee devices is a continuation of the ZHA OTA updates notification idea from related zha-toolkit discussion here -> https://github.com/mdeweerd/zha-toolkit/issues/24 in regards to `ota_notify` in feature zigpy https://github.com/mdeweerd/zha-toolkit#ota_notify (also related to mdeweerd's script https://github.com/mdeweerd/zha-toolkit/blob/7ed030fa51ea18c79e2e592d2252695cc528d42f/examples/fetchOTAfw.sh which download additional Zigbee OTA images from https://github.com/Koenkk/zigbee-OTA).

While the [ZHA integration](https://www.home-assistant.io/integrations/zha) already does support Zigbee OTA updates it currently simply applies all available Zigbee OTA updates as soon as it finds any OTAU files are available, without any UI options for individual Zigbee devices. 

So firstly, would it be possible for ZHA users to only get notifications about Zigbee OTA updates and have UI options to update per device?

Secondly, would it be possible for  ZHA users to get UI options to just select which specific Zigbee devices they would like to OTA update?

Could [ZHA integration](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha) and its UI frontend utilize Home Assistant's new `update` entity type to notify about ZHA OTA updates availability?

https://www.home-assistant.io/integrations/update

https://www.home-assistant.io/integrations/#updates

What does ZHA component + Home Assistant core and frontend UI developers think?

(I guess that if something like that was implemented then it could be reused for Z-Wave JS as well as possibly Matter/Thread in the future?).

Read news in Home Assistant 2022.4 release that introduced  `update`  as a brand new entity type which seams made for features like these?

https://www.home-assistant.io/blog/2022/04/06/release-20224/#introducing-update-entities

_You can now upgrade your devices from inside of Home Assistant just as you would for Home Assistant Releases. You will also notice the number of updates pending in the sidebar by the indicator badge on your configuration cog icon. It's now possible to read the release notes, skip or even install the upgrade to your device in one convenient place._

### _Introducing update entities_

_Say “hi!” to a brand new entity type: update._

_Update entities can tell you if an update is available for your device and service and, in some cases, allow you to install the update straight from Home Assistant!_

_And the beautiful thing is: They show up in your configuration Dashboard, just like a Home Assistant, Home Assistant OS, or add-on update. As a matter of fact, those are now update entities too!_

_Now updates are entities, other integrations can provide them too! For example, a WLED firmware update for your LED strip will show up and installs with a single click._

https://www.home-assistant.io/images/blog/2022-04/update-install.gif

https://building.open-home.io/content/images/2022/04/344c1e8be0c066e49cd4aa8b639d1b50.gif

_Screen recording showing a WLED firmware update can be installed with a clickA WLED firmware update can now be installed just like any other update._

_These `update` entities can be provided by integrations and provide more information about the update, like: Version information, links to release notes, and the possibility to skip that specific version offered._

_The blue number indicator on the configuration cog icon in your sidebar tells you how many updates are pending._

_The Home Assistant Supervisor (providing Core, OS, and add-on updates), WLED, Pi-hole, and Synology DSM integrations have implemented these brand new update entities in this release._

_And because they are all entities, it means you can automate with them. For example, you could send a notification when a new add-on update is available for installation._

**Comment by Hedda** [comment]
FYI, in related news Home Assistant's Z-Wave (Z-Wave JS) developers exposed OTA updates for Z-Wave devices in UI from Home Assistant 2022.7 release, and they also created a new Firmware Update Service (web service with an API) to power semi-automatic firmware updates: 

https://github.com/zwave-js/firmware-updates

https://github.com/zwave-js/firmware-updates/blob/main/README.md

The firmware update service for Z-Wave JS

https://firmware.zwave-js.io

They talk more about it in this video and blog post:

https://youtu.be/HnA_pZQyZAs?t=2908

https://www.home-assistant.io/blog/2022/07/06/release-20227/#update-z-wave-devices-directly-from-home-assistant

## **_Z-Wave JS Firmware Update Service_**

**_This is the web service powering semi-automatic firmware updates in Z-Wave JS._**

### **_How it works_**

**_Firmware upgrades are defined in the files in the firmware directory. This information is parsed by the web service and used to respond to requests._**

**_Z-Wave JS will query the web service on demand, providing the device identification of the device to be updated. If an update is available, the web service will respond with the necessary information to display to the user._**

**_When desired, Z-Wave JS will download the update(s) from the provided URL, verify their integrity and install them_**

## **_Update Z-Wave devices directly from Home Assistant_**

**_All software has bugs, including the software on your Z-Wave devices. But how to update those? Have no fear! As of today, we have a solution for this!_**

**_Thanks to the hard work and collaboration between [@AlCalzone](https://github.com/AlCalzone) from Z-Wave JS and [@raman325](https://github.com/raman325) from Home Assistant, you can now install updates on your Z-Wave devices directly from the Home Assistant interface!_**

**_Screenshot showing the new upload firmware item on the device page of a Z-Wave device page_**

![image](https://user-images.githubusercontent.com/6320001/178096877-c1deac6a-a0a9-4f00-a438-0ebde8d42fb5.png)

**_On the device page of your Z-Wave device, there is now a menu item allowing you to install Z-Wave firmware updates onto your device manually._**

**_Get a software update for your Z-Wave device from the manufacturer, and start an update in Home Assistant, which allows you to upload the update file you got from the manufacturer. The rest is pure magic!_**


**_Screenshot showing the a Z-Wave device that has firmware installation in progress_**

![image](https://user-images.githubusercontent.com/6320001/178096860-5c115502-5bae-4721-9264-bc3942379411.png)

**_During update installation, you’ll be presented with the update’s progress._**

**_It is that easy_** 💪

PS: They also made a tool is meant to help generate the integrity hash for that Z-Wave JS firmware update service:

https://github.com/zwave-js/firmware-integrity

**Comment by Hedda** [comment]
FYI, noticed that the new Z-Wave JS Firmware Update Service (and update server) is now available in UI as of Home Assistant 2022.9 release:

[https://www.home-assistant.io/blog/2022/09/07/release-20229/#z-wave-firmware-updates-now-live](https://www.home-assistant.io/blog/2022/09/07/release-20229/#z-wave-firmware-updates-now-live)

### *Z-Wave Firmware updates now live*

![image](https://user-images.githubusercontent.com/6320001/189090761-4ae2b914-24de-47cf-b573-a9da80b9b606.png)

***In 2022.7, we added support for updating Z-Wave device firmware, but the catch was that you needed the firmware files from the manufacturer in order to complete the update. [@AlCalzone](https://github.com/alcalzone) wasn’t satisfied with this approach and began working on a better solution. That better solution is here with the introduction of the Z-Wave JS Firmware Update Service! For supported manufacturers, Z-Wave JS can now automatically detect, download, and install a firmware update if it is available for your device. This will be done the same way you are currently updating Home Assistant. You will see a notification in your settings menu and you can update right from there. Thanks [@raman325](https://github.com/raman325)***

***For now, only Jasco products are supported. Nabu Casa has been talking to other manufacturers to get more devices supported. To ensure our users can take advantage of this feature, we added update entities for your devices that will check once a day for new firmware updates. Nifty!***

https://www.youtube.com/watch?v=wmYocYd0wts&t=390s&ab_channel=EverythingSmartHome

**Comment by Hedda** [comment]
FYI, OpenHAB's Zigbee developers and users are in an OpenHAB forum thread also discuss GUI/UI management features for such Zigbee OTA firmware updates in the existing unified firmware management system in OpenHAB 3.x / OH3 which apparently already existed in OpenHAB 2.x / OH2) versions but that not yet been migrated to to the web-GUI of OpenHAB 3.x as mentioned here:

https://community.openhab.org/t/ikea-tradfri-firmware-updates/132042/67

and they have this open issue for tracking this missing feature in newer OpenHAB webui:

https://github.com/openhab/openhab-webui/issues/1427

> Firmware update management is not implemented in OH3:  The firmware API, which reads the list of available firmware for a thing, and allows the user to select a firmware, and then provides the status information when firmware is loading is not implemented in OH3. This is a core feature of OH so probably should have support included in the UI as it was in OH2.

The developer of [OpenHAB's ZigBee Binding](https://www.openhab.org/addons/bindings/zigbee/) (who is also the developer of [com.zsmartsystems.zigbee](https://github.com/zsmartsystems/com.zsmartsystems.zigbee)  https://github.com/zsmartsystems/com.zsmartsystems.zigbee/ Zigbee Cluster Library Java framework/libraries it depends on) also mention this update management system in his PR for getting Zigbee OTA firmware images from GitHub repository/repositories:

https://github.com/openhab/org.openhab.binding.zigbee/pull/767

> This PR implements a `FirmwareProvider` for Zigbee devices. It gets the firmware files from the web and caches them locally for use based on requests from the device. The provider gets the firmware IDs from the `ZigBeeThingHandler` based on these periodic requests from the device, and provides them as a list so the user can allow the upgrade via the standard OH firmware management system.
> 
> Requires an updated ZSS `1.4.7` library to provide callback when the client sends a request. This is a little cleaner than the other approach of adding a `ZclCommandListener` as it better integrates in with the server. Also requires further testing

PS: Off-topic but note there that same OpenHAB Zigbee developer is apparently also working on a GitHub Zigbee Firmware Provider backed by GitHub repository/repositories, using a GitHub Library Reader to parse "index.json" and will by default point to Koenkk's https://github.com/Koenkk/zigbee-OTA repo (that is also used by Zigbee2MQTT and IoBroker via zigbee-herdsman), which is also something that has been discussed for zigpy, see -> https://github.com/zigpy/zigpy/issues/535

**Reply by MattWestb** [reply]
I think hosting OTA files is now go for ZHA then its more then likely getting sued from some large companys that we are distributing there code without have one written agreement doing it.
Linking to "repros" is one other thing but hosting is not.

**Reply by Hedda** [reply]
I guess that zigpy could otherwise just have a generic ”GitHub Repository OTA Provider” with a parser where the ZHA user would only have to input the URL to a GitHub repository themselves in the UI could maybe work as long as that URL points to a GitHub repo that contains a correctly formatted ”index.json” file and compatible Zigbee OTA image files?

 Then the user could have the option to choose to simply input https://github.com/Koenkk/zigbee-OTA as URL or enter URLs for some other GitHub repositories?

**Comment by Hedda** [comment]
FYI, [pipiche38](https://github.com/pipiche38) looks to be contemplating implementing similar GUI/UI notifications and management for OTA in Zigbee Plugin for Domoticz:

https://github.com/zigbeefordomoticz/Domoticz-Zigbee/issues/1293

https://github.com/zigbeefordomoticz/Domoticz-Zigbee/pull/1294

**Comment by KaherdinTristan** [comment]
This would be awesome!

**Comment by hwzolin** [comment]
@Hedda Hello, I added the device through ZHA. However, there is still no "update" related operation in the device details. In addition to downloading OTA files to my local computer, what else can I do to get my device OTA?

**Reply by Hedda** [reply]
> Hello, I added the device through ZHA. However, there is still no "update" related operation in the device details. In addition to downloading OTA files to my local computer, what else can I do to get my device OTA?

The original posts above are at this point only a feature request and open discussion/question to [ZHA integration](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha) developers, since while Home Assistant core does now have this new "`update`" entity type that function is not yet utilised by the [ZHA integration](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha) or UI/GUI frontend.  Some other integrations for Home Assistant such as the Z-Wave integration have however already started utilizing it.

https://community.home-assistant.io/t/zha-integration-ui-options-to-apply-ota-firmware-update-otau-per-zigbee-device/438227/

https://www.home-assistant.io/blog/2022/04/06/release-20224/#introducing-update-entities

https://www.home-assistant.io/integrations/update

https://www.home-assistant.io/integrations/#updates

So in summery there is unfortunately not yet any such "update" features related to Zigbee device OTA firmware update functions in the UI/GUI for the [ZHA integration](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha) in Home Assistant, and for now the [ZHA integration](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha) is still limited to just fully automatic updates only without any UI/GUI notification on updates or related UI/GUI configuration settings:

https://www.home-assistant.io/integrations/zha#ota-firmware-updates

https://github.com/zigpy/zigpy/wiki/OTA-Device-Firmware-Updates#enabling-ota-updates

You can today also manually initiating an OTA firmware update using the zha.issue_zigbee_cluster_command service

https://github.com/zigpy/zigpy/wiki/OTA-Device-Firmware-Updates#manually-initiating-ota-firmware-update

Currently, the only OTA update feedback is via the home-assistant.log file

https://github.com/zigpy/zigpy/wiki/OTA-Device-Firmware-Updates#monitoring-ota-upgrades

For now, think best workaround option for ZHA users is to install zha-toolkit and use its ota_notify to manually trigger OTA updates:

https://github.com/mdeweerd/zha-toolkit#ota_notify

PS: I believe that utilizing this new "`update`" entity type for updates notification as well as UI/GUI configuration of OTA firmware update for individual Zigbee device settings might now be the last major missing features from the ZHA integration when comparing it to other popular Zigbee hubs/gateway appliances and applications(?).

**Comment by Hedda** [comment]
FYI, victorigualada wrote a step-by-step guide in the Home Assistant community forum on all the elaborated steps currently needed for an end-user to manually apply a custom Zigbee OTA firmware image to a device in the ZHA integration if there is not an OTA provider in zigpy:

https://community.home-assistant.io/t/zha-ota-firmware-update-upgrade/488960

**Reply by KaherdinTristan** [reply]
Hedda
You seem to be initiated into the world of ZigBee. Do you know of if there is a culture of Tuya releasing fw updates for it's in wall switches? 
And, if yes, where to find them 😂

I am enthusiastic user of ZigBee devices, but the large cluster of Tuya devices on the market makes it difficult for me to avoid buying them, and I would love to be able to update the fw of these two fix hw issues...

💪🏼

Thank you for your participation, contribution and dedication regarding ZigBee, ZHA and Home Assistant!

It does not go unnoticed!
🤓

**Reply by Hedda** [reply]
> Do you know of if there is a culture of Tuya releasing fw updates for it's in wall switches? And, if yes, where to find them 😂
> 
> I am enthusiastic user of ZigBee devices, but the large cluster of Tuya devices on the market makes it difficult for me to avoid buying them, and I would love to be able to update the fw of these two fix hw issues...

Probably better if [start a new dedicated discussion for that question](https://github.com/zigpy/zigpy/discussions) if you want more elaborate answers so this will not go too off-topic.

However, here is a short summary of how I personally see that whole situation today;

Unfortunately, currently, only a few manufacturers are well known to offer public OTA Provider Servers that are accessible sources of official Zigbee device OTAU firmware image files that are usable by independent third parties such as open-source applications for home automation, such as example; Home Assistant, Zigbee2MQTT, IoBroker, OpenHAB, Domoticz, and Jeedom. A few of those projects try to document those we know are hosting official and unofficial Zigbee OTAU firmware image files. Examples:

https://github.com/zigpy/zigpy/blob/dev/zigpy/ota/OTA_URLs.md

https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/lib/ota/OTA_URLs.md

Not all but many manufacturers probably have OTA Provider Servers that are only accessible to closed ecosystems such as proprietary gateways/hubs, such as the example the Philips Hue Bridge(?). Though there are some unofficial sources of Zigbee device OTAU firmware image files that have been extracted in one or another way from Zigbee devices, such as for example Koenkk's collection of Zigbee OTA files:

https://github.com/Koenkk/zigbee-OTA

In my experience best is to try as an end-user/customer to get in contact with the manufacturer of Zigbee devices directly and ask if they already have an OTA provider server or other ways of hosting Zigbee device OTAU firmware image files and convince/request them to make such public and accessible so that they can be used by popular independent third parties such as open-source applications for home automation.

Then there are probably more manufacturers of Zigbee devices that do not provide any OTA firmware updates at all for post-sale install.

As for Tuya specifically, I can only speculate, but I believe that Tuya's business model is primarily around being an [OEM (Original Equipment Manufacturer)](https://en.wikipedia.org/wiki/Original_equipment_manufacturer) and/or [ODM (Original Design Manufacturer)](https://en.wikipedia.org/wiki/Original_design_manufacturer), meaning that I think they mostly make products that will be rebranded and sold by other companies that that their customers and if so then assume that they offer firmware updates to those companies on demand if they specifically pay for that service. Anyway, as you can see there are not many Tuya firmware images available in the wild even though they manufacture loads of different Zigbee devices -> https://github.com/Koenkk/zigbee-OTA/tree/master/images/Tuya

**Comment by Hedda** [comment]
@puddly based on your proof of concept synchronous OTA manager in https://github.com/zigpy/zigpy/pull/1102 sounds as if you have plans for device specific OTA updates?

> **_I'm hoping to move request/response matching into the main ControllerApplication, allowing for more granular control over requests/responses within async code and for dynamically attaching objects that expect responses. In the future, this will allow for sending requests to NWK addresses directly and to wait for a response, without the need for intermediate `Device` objects._**
> 
> **_As a proof of concept, below is a synchronous OTA manager that can send a specific OTA image to a specific device, without modifying the global ZCL cluster definitions  ..._**

**Reply by Hedda** [reply]
Any updates on device specific OTA updates or OTA notify support?

**Comment by mbbush** [comment]
Giving the user more control over when/whether to apply an OTA update seems really valuable, and unambiguously a good thing. It's completely separate from any copyright concerns around redistributing firmware files. I'd love to see this feature, even if it's (at least initially) limited to the handful of manufacturers with existing OTA providers.

**Comment by mkruiver** [comment]
I would still like to have this feature!

**Comment by theorlangur** [comment]
Yes, please
That's a natural functionality that's currently missing

**Comment by Hedda** [comment]
Also see feature request from @pipiche38 asking for ability to trigger OTA directly via zigpy-cli commands here -> https://github.com/zigpy/zigpy-cli/issues/39

I believe such a feature could utilize the recently merged application-level request/response matching -> https://github.com/zigpy/zigpy/pull/1102

(Note again that puddly posted in that PR some proof of concept code for sending a specific OTA image to a specific device).

**Comment by mhl66** [comment]
The current OTA update of IKEA messed up the the whole setup, see https://community.home-assistant.io/t/ikea-tradfri-shortcut-button-battery-shows-255-percent-intead-of-100-percent/632023 - and, no group pairing possible for STYRBAR (E2001/2002) and TRADFRI anymore.

**Reply by Hedda** [reply]
> The current OTA update of IKEA messed up the the whole setup, see https://community.home-assistant.io/t/ikea-tradfri-shortcut-button-battery-shows-255-percent-intead-of-100-percent/632023 - and, no group pairing possible for STYRBAR (E2001/2002) and TRADFRI anymore.

If re-pairing the device does not help then suggest that you report that as an ZHA issue with Home Assistant Core (home assistant's core repository on GitHub) if no one else already opened a new issue for it there, see -> https://www.home-assistant.io/integrations/zha#reporting-issues -> https://github.com/home-assistant/core/issues

**Reply by MattWestb** [reply]
Its very known that IKEA have taking away group **binding** in  the latest firmware also for first and second gen controllers for over one year ago so its expected.
Also Styrbar is the first controller with battery reading fix that is not working well but on all newer releases its working OK and is well documented in the git issues.

**Reply by puddly** [reply]
@mhl66 Please moderate your tone. Remember that you're speaking to contributors who work on this project in their free time.

**Reply by mhl66** [reply]
Sorry, did not mean any insult.

**Reply by Hedda** [reply]
Regardless, it is very much off-topic fom the feature request discussion that this thread is about.

**Reply by mhl66** [reply]
True.

**Comment by ssorgatem** [comment]
This is one of the things missing from ZHA that if implemented, IMO, would make it definitely superior to Z2M.

**Comment by Hedda** [comment]
FYI, dmulcahey submitted pull requests with code patches for zigpy OTA handling + update platform to ZHA integration component, see:

https://github.com/home-assistant/core/pull/107612

and

https://github.com/zigpy/zigpy/pull/1321

As requested this should add initial support for Home Assistant’s universal update platform and expose Zigbee OTA updates for devices in the same unified way as other updates are in Hom

https://www.home-assistant.io/blog/2022/04/06/release-20224/#introducing-update-entities

Backstory if you missed it; Home Assistant core has that relatively new unified update platform and “update” entity type that can be initiated from the same place from UI/GUI by end users:

https://www.home-assistant.io/integrations/update

Again, several other integrations for Home Assistant, (such as the similar Z-Wave integration), are already utilizing thos new update platform to add update entities UI for device OTAU updates:

https://www.home-assistant.io/integrations/#updates

"_Now updates are entities, other integrations can provide them too! For example, a WLED firmware update for your LED strip will show up and installs with a single click._"

https://www.home-assistant.io/images/blog/2022-04/update-install.gif

"_These `update` entities can be provided by integrations and provide more information about the update, like: Version information, links to release notes, and the possibility to skip that specific version offered._"

More references:

https://www.home-assistant.io/blog/2022/09/07/release-20229/#z-wave-firmware-updates-now-live

https://www.youtube.com/watch?v=wmYocYd0wts&t=390s&ab_channel=EverythingSmartHome

https://www.home-assistant.io/blog/2022/07/06/release-20227/#update-z-wave-devices-directly-from-home-assistant

**Comment by Hedda** [comment]
FYI, the pull requests by dmulcahey has now been merged so this should be in Home Assistant 2024.2 (and is already in the 2024.2 beta):

https://github.com/home-assistant/core/pull/107612

https://github.com/zigpy/zigpy/pull/1321

https://www.home-assistant.io/blog/2024/02/07/release-20242/#updating-your-zigbee-devices

## ***Update entities for ZHA devices***

***[ZHA](https://rc.home-assistant.io/integrations/zha) (Zigbee Home Automation, our built-in Zigbee integration) now provides update entities to Home Assistant!***

***This means you can get notified when a device has a firmware update available, and you can trigger the update from Home Assistant!***

![image](https://github.com/zigpy/zigpy/assets/6320001/b78b882b-9453-462b-8c68-503f44ee0003)

***Noticed the weird version numbers? Unfortunately, that is how version numbers in the Zigbee standard work and look, we currently have no way to translate them to something more human-readable format.***

***Currently, it supports updating Inovelli, OSRAM/Ledvance, Sonoff, and ThirdReality devices, and we are looking at adding support for more brands in the future.***
